File: src/morie/fn/fwpas.py
# morie.fn — function file (hadesllm/morie)
"""Dense layer forward pass."""
from __future__ import annotations

import numpy as np

from ._richresult import RichResult

__all__ = ["forward_pass_dense"]


def _apply_activation(z: np.ndarray, activation: str) -> np.ndarray:
    a = activation.lower()
    if a in ("identity", "linear", "none"):
        return z
    if a == "sigmoid":
        # Numerically stable sigmoid
        out = np.empty_like(z)
        pos = z >= 0
        out[pos] = 1.0 / (1.0 + np.exp(-z[pos]))
        ez = np.exp(z[~pos])
        out[~pos] = ez / (1.0 + ez)
        return out
    if a == "tanh":
        return np.tanh(z)
    if a == "relu":
        return np.maximum(0.0, z)
    if a == "softmax":
        zs = z - np.max(z, axis=-1, keepdims=True)
        e = np.exp(zs)
        return e / np.sum(e, axis=-1, keepdims=True)
    raise ValueError(f"Unknown activation: {activation!r}")


def forward_pass_dense(x, w, b, activation: str = "sigmoid"):
    r"""Dense layer forward pass.

    Computes :math:`z = W x + b` followed by an element-wise
    activation :math:`a = \\sigma(z)`.

    Parameters
    ----------
    x : array-like, shape ``(n_in,)`` or ``(batch, n_in)``
        Input data (single sample or mini-batch).
    w : array-like, shape ``(n_out, n_in)``
        Weight matrix.
    b : array-like, shape ``(n_out,)``
        Bias vector.
    activation : str
        ``'sigmoid'`` (default), ``'tanh'``, ``'relu'``, ``'softmax'``, or
        ``'identity'``.

    Returns
    -------
    result : RichResult
        Keys: ``z`` (pre-activation), ``a`` / ``estimate`` (post-activation).

    References
    ----------
    Goodfellow, Bengio & Courville (2016) *Deep Learning*, Ch 6.
    """
    x = np.asarray(x, dtype=float)
    w = np.asarray(w, dtype=float)
    b = np.asarray(b, dtype=float)
    # x @ W^T + b  (broadcasts over batch dim if present)
    z = x @ w.T + b
    a = _apply_activation(z, activation)
    return RichResult(
        title="Dense layer forward pass",
        summary_lines=[("Activation", activation), ("Output shape", a.shape)],
        payload={
            "z": z,
            "a": a,
            "estimate": a,
            "activation": activation,
            "method": "Dense layer forward pass",
        },
    )


# CANONICAL TEST
# x=[1,2], W=[[1,0],[0,1],[1,1]], b=[0,0,0], identity -> z=a=[1,2,3]
# softmax of [1,2,3] -> [0.0900, 0.2447, 0.6652]


def cheatsheet():
    return "fwpas: Dense layer forward pass z=Wx+b, a=sigma(z)"

File: r-package/morie/R/fwpas.R
# SPDX-License-Identifier: MIT OR Apache-2.0

#' Dense layer forward pass
#'
#' Pure-base-R parity for \code{morie.fn.fwpas.forward_pass_dense}.
#'
#' \deqn{z = W x + b, \quad a = \sigma(z)}
#'
#' @param x Numeric vector (single input) or matrix (rows = samples).
#' @param w Weight matrix shape \code{(n_out, n_in)}.
#' @param b Bias vector length \code{n_out}.
#' @param activation One of \code{"sigmoid"}, \code{"tanh"},
#'   \code{"relu"}, \code{"softmax"}, \code{"identity"}.
#' @return Named list with components \code{z}, \code{a}, \code{estimate}
#'   (= \code{a}), \code{activation}, \code{method}.
#' @references Goodfellow, Bengio & Courville (2016), Deep Learning, Ch 6.
#' @examples
#' \dontrun{
#'   # See the package vignettes for usage examples:
#'   #   vignette(package = "morie")
#' }
#' @export
fwpas_forward_pass_dense <- function(x, w, b, activation = "sigmoid") {
  x <- as.matrix(x)
  w <- as.matrix(w)
  b <- as.numeric(b)
  if (ncol(x) == 1L && nrow(x) == length(b) + (ncol(w) - 1L)) {
    # 1-D input passed as column vector
    x <- t(x)
  } else if (ncol(x) != ncol(w)) {
    # try transpose if user passed (n_in,)
    if (nrow(x) == ncol(w)) x <- t(x)
  }
  z <- x %*% t(w)
  z <- sweep(z, 2L, b, "+")
  a <- switch(activation,
    "identity" = z,
    "linear"   = z,
    "none"     = z,
    "sigmoid"  = 1 / (1 + exp(-z)),
    "tanh"     = tanh(z),
    "relu"     = pmax(0, z),
    "softmax"  = {
      ez <- exp(z - apply(z, 1L, max))
      sweep(ez, 1L, rowSums(ez), "/")
    },
    stop(sprintf("Unknown activation: %s", activation))
  )
  list(z = z, a = a, estimate = a, activation = activation,
       method = "Dense layer forward pass")
}

#' @rdname fwpas_forward_pass_dense
#' @keywords internal
#' @export
forward_pass_dense <- fwpas_forward_pass_dense
